File: 29.js
function permuteWithBlocked(nums, blocked) {
    const res = [];
    const free = [];
    const fixed = [...nums];

    for (let i = 0; i < nums.length; i++) {
        if (!blocked[i]) free.push(nums[i]);
    }

    free.sort((a, b) => a - b);
    const used = Array(free.length).fill(false);

    function backtrack(path) {
        if (path.length === free.length) {
            let idx = 0;
            const perm = [...fixed];
            for (let i = 0; i < nums.length; i++) {
                if (!blocked[i]) perm[i] = path[idx++];
            }
            res.push(perm);
            return;
        }

        for (let i = 0; i < free.length; i++) {
            if (used[i]) continue;
            if (i > 0 && free[i] === free[i - 1] && !used[i - 1]) continue;
            used[i] = true;
            path.push(free[i]);
            backtrack(path);
            path.pop();
            used[i] = false;
        }
    }

    backtrack([]);
    return res;
}

File: 29.py
def permuteWithBlocked(nums, blocked):
    res = []
    n = len(nums)
    fixed = nums[:]
    free = [nums[i] for i in range(n) if not blocked[i]]
    free.sort()
    used = [False] * len(free)

    def backtrack(path):
        if len(path) == len(free):
            idx = 0
            perm = fixed[:]
            for i in range(n):
                if not blocked[i]:
                    perm[i] = path[idx]
                    idx += 1
            res.append(perm)
            return

        for i in range(len(free)):
            if used[i]: continue
            if i > 0 and free[i] == free[i-1] and not used[i-1]:
                continue
            used[i] = True
            backtrack(path + [free[i]])
            used[i] = False

    backtrack([])
    return res
